text/0131-pass-native-sdk-spans-to-hybrid.md

```diff
@@ -0,0 +1,174 @@
+4. After the hybrid SDKs pull spans from the native SpansAggregator, the native SpansAggregator
+removes all its spans.
+
+```Gherkin
+Scenario: No active span, native SpansAggregator flushes
+    Given no active span
+    When the native SpansAggregator has to flush all its spans
+    Then native SDK pushes all native spans up to the hybrid SDK
+    And doesn't send its native spans directly to Sentry
+    And the hybrid SDK adds all native spans to its SpansAggregator
+
+Scenario: Active span, native SpansAggregator flushes
+    Given an active span
+    When the native SpansAggregator has to flush all its spans
+    Then native SDK pushes all native spans up to the hybrid SDK
+    And the hybrid SDK filters for spans belonging to the active span 
+    And adds these spans to the active span
+    And adds the non matching spans to the hybrid SpansAggregator
+
+Scenario: Active span finishes
+    Given an active span
+    When it finishes
+    Then hybrid SDK gets all spans from the native SpansAggregator
+    And the hybrid SDK filters for spans belonging to the active span 
+    And adds these spans to the active span
+    And adds the non matching spans to the hybrid SpansAggregator
+
+Scenario: Active span, hybrid SpansAggregator flushes
+    Given an active span
+    When the hybrid SpansAggregator has to flush all its spans
+    Then it does not pull spans from the native SpansAggregator
+    And sends all hybrid spans from the SpansAggregator to Sentry
+
+Scenario: No active span, hybrid SpansAggregator flushes
+    Given no active span
+    When the hybrid SpansAggregator has to flush all its spans
+    Then it pulls spans from the native SpansAggregator
+    And adds these to the hybrid SpansAggregator
+    And sends native and hybrid spans to Sentry
+```
+
+### Pros <a name="option-2-pros"></a>
+
+1. Reusing existing functionality of the SpansAggregator.
+2. Works well with single-span ingestion and transactions.
+3. Adding spans to the SpansAggregator is `O(1)`.
+4. Less cross hybrid layer communication than [option 3](#option-3).
+
+### Cons <a name="option-2-cons"></a>
+
+1. Please add your cons.
+
+
+## Option 3: Native Callbacks<a name="option-3"></a>
+
+The native layers offer callbacks whenever they start and finish auto-instrumented spans and don't
+keep track of spans. For example, the HTTP request instrumentation offers two callbacks: one for
+starting the HTTP request and another for finishing the HTTP request. The hybrid SDKs then subscribe
+to these callbacks and create the spans themselves.
+
+### Pros <a name="option-3-pros"></a>
+
+1. This option works well with idle transactions and `waitForChildren` because these concepts need to
+know when spans start and finish, but this is a weak argument as with single-span ingestion, this
+concept will become obsolete.
+2. Works well with single-span ingestion, as the hybrid SDKs get callbacks for each span.
```

Comment:
we recently discussed that to calculate exclusive time in single-span ingestion, all child spans need to be finished. does that mean that this option would provide individual callbacks for each span, but they would likely arrive at the same time because we'd need to calculate exclusive time and that's only possible once all (related) spans finished?

Reply:
As I understand it, this would be only for the top-level spans, if the native DB query has children those would not be exposed.

Reply:
ah that makes sense 👍

Reply:
The auto instrumented spans on Cocoa we are interested in this RFC aren't nested. DB queries, HTTP requests, and file IO operations don't have nesting. We can nest them on the hybrid layer.